src/main/java/org/elasticsearch/index/shard/ShardStateMetaData.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.gateway.CorruptStateException;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.Path;
+import java.util.regex.Pattern;
+
+/**
+ */
+public final class ShardStateMetaData {
+
+    private static final String SHARD_STATE_FILE_PREFIX = "state-";
+    private static final Pattern SHARD_STATE_FILE_PATTERN = Pattern.compile(SHARD_STATE_FILE_PREFIX + "(\\d+)(" + MetaDataStateFormat.STATE_FILE_EXTENSION + ")?");
+    private static final String PRIMARY_KEY = "primary";
+    private static final String VERSION_KEY = "version";
+    private static final String INDEX_UUID_KEY = "index_uuid" ;
+
+    public final long version;
+    public final String indexUUID;
+    public final boolean primary;
+
+    public ShardStateMetaData(long version, boolean primary, String indexUUID) {
+        assert indexUUID != null;
+        this.version = version;
+        this.primary = primary;
+        this.indexUUID = indexUUID;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        ShardStateMetaData that = (ShardStateMetaData) o;
+
+        if (primary != that.primary) return false;
+        if (version != that.version) return false;
+        if (indexUUID != null ? !indexUUID.equals(that.indexUUID) : that.indexUUID != null) return false;
```

Comment:
yes I know but they both have `IndexMetaData#INDEX_UUID_NA_VALUE ` no? I mean we never change the UUID of an index so it should be ok?